src/Workspaces/Core/Portable/Workspace/Solution/ProjectDependencyGraph_RemoveProject.cs

```diff
@@ -0,0 +1,167 @@
+            return builder.ToImmutable();
+        }
+
+        /// <summary>
+        /// Computes a new <see cref="_lazyReverseReferencesMap"/> for the removal of a project.
+        /// </summary>
+        private static ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>>? ComputeNewReverseReferencesMapForRemovedProject(
+            ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> existingForwardReferencesMap,
+            ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>>? existingReverseReferencesMap,
+            ProjectId projectId)
+        {
+            if (existingReverseReferencesMap is null)
+            {
+                // The map was never calculated for the previous graph, so there is nothing to update.
+                return null;
+            }
+
+            var builder = existingReverseReferencesMap.ToBuilder();
+
+            // Iterate over each project referenced by 'projectId', which is now being removed. Update the reverse
+            // references map for the project to no longer include 'projectId' in the list.
+            if (existingForwardReferencesMap.TryGetValue(projectId, out var referencedProjectIds))
+            {
+                foreach (var referencedProjectId in referencedProjectIds)
+                {
+                    if (builder.TryGetValue(referencedProjectId, out var reverseDependencies))
+                    {
+                        builder[referencedProjectId] = reverseDependencies.Remove(projectId);
+                    }
+                }
+            }
+
+            // Finally, remove 'projectId' itself.
+            builder.Remove(projectId);
+            return builder.ToImmutable();
+        }
+
+        /// <summary>
+        /// Computes a new <see cref="_transitiveReferencesMap"/> for the removal of a project.
+        /// </summary>
+        /// <seealso cref="ComputeNewReverseTransitiveReferencesMapForRemovedProject"/>
+        private static ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> ComputeNewTransitiveReferencesMapForRemovedProject(
+            ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> existingTransitiveReferencesMap,
+            ProjectId projectId)
+        {
+            var builder = existingTransitiveReferencesMap.ToBuilder();
+
+            // Iterate over each project and invalidate the transitive references for the project if the project has an
+            // existing transitive reference to 'projectId'.
+            foreach (var (project, references) in existingTransitiveReferencesMap)
+            {
+                if (references.Contains(projectId))
+                {
+                    // The project transitively referenced 'projectId', so any transitive references brought in
+                    // exclusively through this reference are no longer valid. Remove the project from the map and the
+                    // new transitive references will be recomputed the first time they are needed.
```

Comment:
oh that's interesting.  have we doc'ed that these maps are 'partial'?  i.e. that even if they're non-null, they may not have the information about all projects and may be faulted in.  That def wasn't something i was aware of while reading this up till now.

Reply:
Yes, they are documented.